src/controls/control.class.ts

```diff
@@ -1,372 +1,395 @@
-          fabric.util.degreesToRadians(objectAngle);
-        cosHalfOffset = cornerHypotenuse * fabric.util.cos(newTheta);
-        sinHalfOffset = cornerHypotenuse * fabric.util.sin(newTheta);
-        // use complementary angle for two corners
-        cosHalfOffsetComp = cornerHypotenuse * fabric.util.cos(newThetaComp);
-        sinHalfOffsetComp = cornerHypotenuse * fabric.util.sin(newThetaComp);
-      } else {
-        // handle square corners
-        // use default object corner size unless size is defined
-        var cornerSize = xSize && ySize ? xSize : objectCornerSize;
-        /* 0.7071067812 stands for sqrt(2)/2 */
-        cornerHypotenuse = cornerSize * 0.7071067812;
-        // complementary angles are equal since they're both 45 degrees
-        var newTheta = fabric.util.degreesToRadians(45 - objectAngle);
-        cosHalfOffset = cosHalfOffsetComp =
-          cornerHypotenuse * fabric.util.cos(newTheta);
-        sinHalfOffset = sinHalfOffsetComp =
-          cornerHypotenuse * fabric.util.sin(newTheta);
-      }
+    return {
+      tl: new Point(centerX - sinHalfOffsetComp, centerY - cosHalfOffsetComp),
+      tr: new Point(centerX + cosHalfOffset, centerY - sinHalfOffset),
+      bl: new Point(centerX - cosHalfOffset, centerY + sinHalfOffset),
+      br: new Point(centerX + sinHalfOffsetComp, centerY + cosHalfOffsetComp),
+    };
+  }
 
-      return {
-        tl: {
-          x: centerX - sinHalfOffsetComp,
-          y: centerY - cosHalfOffsetComp,
-        },
-        tr: {
-          x: centerX + cosHalfOffset,
-          y: centerY - sinHalfOffset,
-        },
-        bl: {
-          x: centerX - cosHalfOffset,
-          y: centerY + sinHalfOffset,
-        },
-        br: {
-          x: centerX + sinHalfOffsetComp,
-          y: centerY + cosHalfOffsetComp,
-        },
-      };
-    },
+  /**
+   * Render function for the control.
+   * When this function runs the context is unscaled. unrotate. Just retina scaled.
+   * all the functions will have to translate to the point left,top before starting Drawing
+   * if they want to draw a control where the position is detected.
+   * left and top are the result of the positionHandler function
+   * @param {RenderingContext2D} ctx the context where the control will be drawn
+   * @param {Number} left position of the canvas where we are about to render the control.
+   * @param {Number} top position of the canvas where we are about to render the control.
+   * @param {Object} styleOverride
+   * @param {FabricObject} fabricObject the object where the control is about to be rendered
+   */
+  render(
+    ctx: CanvasRenderingContext2D,
+    left: number,
+    top: number,
+    styleOverride: ControlRenderingStyleOverride | undefined,
+    fabricObject: FabricObject
+  ) {
+    styleOverride = styleOverride || {};
+    switch (styleOverride.cornerStyle || fabricObject.cornerStyle) {
+      case 'circle':
+        renderCircleControl.call(
```

Comment:
I want to make the controls renders bound to the class as well one day

Reply:
The render function isn't bound just because i wanted to handle the old cornerstyle and keep 2 functions separated. `render` could have been just another option to add to the controls deprecating the `cornerStyle ` option in the  object.

It would be more logic to have a predefined style for controls without any type/corner/color and all the rest is done on the control itself.

Reply:
Feels like you didn't answer why not bound but why separated.
I suggest adding them as private methods as is.

Reply:
seems to me they fall in the exact pattern of a class member